dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java

```diff
@@ -405,4 +408,76 @@ public static long countFailures(List<ArtifactResults> table, String columnName)
         .filter(row -> row.getResult(columnName) == null || row.getFailureCount(columnName) > 0)
         .count();
   }
+
+  private static final int MINIMUM_NUMBER_DEPENDENCY_PATHS = 5;
+
+  /**
+   * Returns mapping from jar files to summaries of the root problem in their {@link
+   * DependencyPath}s. The summary explains common patterns ({@code groupId:artifactId}) in the path
+   * elements. The returned map does not have a key for a jar file when it has fewer than {@link
+   * #MINIMUM_NUMBER_DEPENDENCY_PATHS} dependency paths or a common pattern is not found among the
+   * elements in the paths.
+   *
+   * <p>Example summary: "Artifacts 'com.google.http-client:google-http-client &gt;
+   * commons-logging:commons-logging &gt; log4j:log4j' exist in all 994 dependency paths. Example
+   * path: com.google.cloud:google-cloud-core:1.59.0 ..."
+   *
+   * <p>Using this summary in the BOM dashboard avoids repetitive items in the {@link
+   * DependencyPath} list that share the same root problem caused by widely-used libraries, for
+   * example, {@code commons-logging:commons-logging}, {@code
+   * com.google.http-client:google-http-client} and {@code log4j:log4j}.
+   */
+  private static ImmutableMap<String, String> findRootCauses(
+      ListMultimap<Path, DependencyPath> jarToDependencyPaths) {
+    // Freemarker is not good at handling non-string key. Path object in .ftl is automatically
+    // converted to String. https://freemarker.apache.org/docs/app_faq.html#faq_nonstring_keys
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+
+    for (Path jar : jarToDependencyPaths.keySet()) {
+      List<DependencyPath> dependencyPaths = jarToDependencyPaths.get(jar);
+
+      Set<String> commonVersionlessArtifacts = commonVersionlessArtifacts(dependencyPaths);
+
+      // When dependencyPaths is not empty, versionlessCoordinates is not null
+      if (dependencyPaths.size() > MINIMUM_NUMBER_DEPENDENCY_PATHS
+          && commonVersionlessArtifacts.size() > 1) { // The last paths elements are always same
+        builder.put(
+            jar.toString(),
+            summaryMessage(
+                dependencyPaths.size(), commonVersionlessArtifacts, dependencyPaths.get(0)));
+      }
+    }
+    return builder.build();
+  }
+
+  private static Set<String> commonVersionlessArtifacts(List<DependencyPath> dependencyPaths) {
+    Set<String> versionlessCoordinatesIntersection = null; // null for 1st iteration
```

Comment:
I think you can do something like

ImmutableList<String> v =  dependencyPaths.get(0).getPath().stream().map(Artifacts::makeKey).collect(toImmutableList());
Set<String> versionlessCoordinatesIntersection = Sets.newLinkedHashSet(v);

Then you don't need the special casing for the first iteration of the for loop since you'll just be intersecting with itself. 

You could pull dependencyPath.getPath().stream().map(Artifacts::makeKey).collect(toImmutableList()) out to a helper function to make this clearer and avoid duplication.